src/wallpaper.c

```diff
@@ -0,0 +1,290 @@
+
+typedef struct {
+  XdpImplWallpaper *impl;
+  GDBusMethodInvocation *invocation;
+  Request *request;
+  GtkWidget *dialog;
+  ExternalWindow *external_parent;
+
+  guint response;
+  gchar *picture_uri;
+  SetWallpaperOn set_on;
+} WallpaperDialogHandle;
+
+static void
+wallpaper_dialog_handle_free (gpointer data)
+{
+  WallpaperDialogHandle *handle = data;
+
+  g_clear_object (&handle->external_parent);
+  g_clear_object (&handle->request);
+  g_clear_pointer (&handle->picture_uri, g_free);
+
+  if (handle->dialog != NULL)
+      gtk_widget_destroy (GTK_WIDGET (handle->dialog));
+  g_clear_object (&handle->dialog);
+
+
+  g_free (handle);
+}
+
+static void
+send_response (WallpaperDialogHandle *handle)
+{
+  if (handle->request->exported)
+    request_unexport (handle->request);
+
+  xdp_impl_wallpaper_complete_set_wallpaper_uri (handle->impl,
+                                                 handle->invocation,
+                                                 handle->response);
+
+  wallpaper_dialog_handle_free (handle);
+}
+
+static gboolean
+set_gsettings (gchar *schema,
+               gchar *uri)
+{
+  g_autoptr(GSettings) settings = NULL;
+
+  settings = g_settings_new (schema);
+
+  return g_settings_set_string (settings,
+                                "picture-uri",
+                                uri);
+}
+
+static void
+on_file_copy_cb (GObject *source_object,
+                 GAsyncResult *result,
+                 gpointer data)
+{
+  WallpaperDialogHandle *handle = data;
+  GFile *picture_file = G_FILE (source_object);
+  g_autoptr(GError) error = NULL;
+  g_autofree gchar *uri = NULL;
+
+  handle->response = 2;
+
+  uri = g_file_get_uri (picture_file);
+  if (!g_file_copy_finish (picture_file, result, &error))
+    {
+      g_warning ("Failed to copy '%s': %s", uri, error->message);
+
+      goto out;
+    }
+
+  switch (handle->set_on)
+    {
+      case BACKGROUND:
+        if (set_gsettings (BACKGROUND_SCHEMA, uri))
+          handle->response = 0;
+        break;
+      case LOCKSCREEN:
+        if (set_gsettings (LOCKSCREEN_SCHEMA, uri))
+          handle->response = 0;
+        break;
+      default:
+        if (set_gsettings (BACKGROUND_SCHEMA, uri) &&
+            set_gsettings (LOCKSCREEN_SCHEMA, uri))
+          handle->response = 0;
+    }
+
+out:
+  send_response (handle);
+}
+
+static void
+set_wallpaper (WallpaperDialogHandle *handle)
+{
+  g_autofree gchar *dest_filename = NULL;
+  g_autofree gchar *dest_path = NULL;
+  g_autoptr(GFile) source = NULL;
+  g_autoptr(GFile) destination = NULL;
+
+  switch (handle->set_on)
+    {
+      case BACKGROUND:
+        dest_filename = g_strdup ("background");
+        break;
+      case LOCKSCREEN:
+        dest_filename = g_strdup ("lockscreen");
+        break;
+      default:
+        dest_filename = g_strdup ("both");
+    }
+
+  dest_path = g_build_filename (g_get_user_config_dir (), dest_filename, NULL);
```

Comment:
Was it really intended to copy the chosen image to `~/.config/background` directly as a file, instead of creating a directory by that name and putting the image inside it?

I am asking because in the past Settings used `~/.config/gnome-control-center/backgrounds/...`, and I am not sure if there is a typo somewhere or this was indeed intentional.

Reply:
Ok, I see.  Thanks for clarifying that.

That part of Settings' Background panel predates my involvement with that code.  I think the main reason was to namespace everything related to backgrounds in a single sub-directory.  Other than caching remote wallpapers, like photos from Flickr, it was used to hold a INI-style configuration file that stored the custom colours that one could add to the Background panel, and we used to have some bits of XML too (see `CcBackgroundXml`).  There might have been other things, but these are the ones I could dig up from Git.

Anyway, no stress.  I don't really have an opinion.  A sub-directory might be good for namespacing, but if you already had good reasons for the current approach, then that's fine too.  :)